core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -1036,8 +1006,13 @@ private[ui] class TaskDataSource(
       shuffleRead,
       shuffleWrite,
       bytesSpilled,
-      taskData.errorMessage.getOrElse(""),
-      logs)
+      info.errorMessage.getOrElse(""),
+      executorLogs(info.executorId))
+  }
+
+  private def executorLogs(id: String): Map[String, String] = {
+    executors.getOrElseUpdate(id,
+      store.executorSummary(id).map(_.executorLogs).getOrElse(Map.empty))
```

Comment:
I think you need to protect `executors` from a race if two UI threads both call this.

Reply:
Pretty sure there's a separate instance of this class per request.

Reply:
ah right, good point